### PR TITLE
fix(ci): restore perl-ci-hygiene compilation on master (#10107)

### LINKED ISSUE #10107: master does not compile: perl-ci-hygiene defines is_rust_doctest_fence twice
## Problem

`master` has not compiled since 4f0de95 (2026-08-11).

```
$ cargo check -p perl-ci-hygiene --all-targets --locked
error[E0428]: the name `is_rust_doctest_fence` is defined multiple times
    --> crates/perl-ci-hygiene/src/main.rs:2725:1
     |
2689 | fn is_rust_doctest_fence(fence: &str) -> bool {
     | --------------------------------------------- previous definition here
...
2725 | fn is_rust_doctest_fence(fence: &str) -> bool {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ redefined here
```

## How it landed

PR #10024 (`fix(ci): ratchet test-source panic usage (#4904)`, merged 2026-08-11) added **both** definitions in the same diff — two drafting passes of the same helper, neither removed. The PR merged while its CI run was cancelled, so E0428 never surfaced on a check.

Five commits have landed on `master` since, all on top of a tree that does not build.

The same hunk also carries two mechanical gate failures that CI never reported:

- rustfmt line wrapping in `scan_doctest_source` and `rustdoc_line`
- `clippy::question_mark` (error under `-D warnings`) on the `let ... else { return None }` in `rustdoc_line`

## Which implementation is correct

The two differ in how they split the fence info string:

| | separator | classifies `rust,ignore` |
| --- | --- | --- |
| line 2689 | whitespace | **not Rust** — skipped |
| line 2725 | comma | Rust — scanned |

The repository's own fences settle it. Of 1,022 fenced blocks under `crates/` and `xtask/`:

```
556  (bare)
165  rust,ignore
126  rust
 47  text
 46  ignore
 34  bash
 19  no_run
 17  perl
 10  json
  2  rust,no_run
  2  lua
```

The whitespace-splitting variant reads `rust,ignore` as a single token, matches nothing, and silently skips all 167 comma-form Rust doctests. The comma-splitting variant classifies every form present in the tree correctly and still rejects `text`, `bash`, `perl`, `json`, and `lua`.

Measured against the committed ratchet baselines, both pass; the comma-splitting one detects 5 more real hits:

```
comma-split       1,342 unwrap/expect   550 panic-family
whitespace-split  1,337 unwrap/expect   550 panic-family
baselines         1,528 unwrap/expect   568 panic-family
```

The existing test `doctest_scanner_only_reports_rust_fences` uses only ` ```rust ` and ` ```text `, which both variants classify identically — so it never discriminated between them.

## Scope

- keep the comma-splitting definition, delete the whitespace-splitting one
- fix the rustfmt and `clippy::question_mark` failures in the same block

## Non-goals

- no change to the ratchet baselines (both variants pass under them)
- no change to which fence attributes are recognised beyond removing the duplicate
- nothing in `perl-lsp-swarm` — the development repository modularised `perl-ci-hygiene` into `src/commands/panic_test/` and carries no copy of this code, so there is nothing to sync

## Acceptance

- [ ] `cargo check -p perl-ci-hygiene --all-targets --locked` passes
- [ ] `cargo fmt -p perl-ci-hygiene -- --check` clean
- [ ] `cargo clippy -p perl-ci-hygiene --all-targets --locked -- -D warnings` clean
- [ ] `cargo test -p perl-ci-hygiene --all-targets --locked` passes
- [ ] `perl-ci-hygiene check-unwraps-tests` stays under baseline

## Related

The merge-with-cancelled-CI path is the same failure mode tracked in #7278.

### PR DESCRIPTION
**Pull request by EffortlessSteven**
_Requested by **Steven** · [project thread](https://claude.ai/code/project/chan_016ahUT2GL29gFQZf2VJzJep?thread=cmsg_016ahUT2GL29gFQZf2VJzJepD4GKcP9xaVJb97vo1e8qk6)_

**Before:** `master` does not compile. `cargo check -p perl-ci-hygiene --all-targets --locked` fails with `error[E0428]: the name is_rust_doctest_fence is defined multiple times`. It has been in this state since 4f0de95 on 2026-08-11, and five commits have landed on top of a tree that does not build.

**After:** the crate compiles, formats, lints under `-D warnings`, and its tests pass. `cargo check --workspace --all-targets --locked` also completes clean, so this one crate was the whole of what stopped the tree building. The doctest scanner now classifies fences the way rustdoc does, and the test-source ratchet runs with its counts unchanged.

## Objective

Restore compilation of `perl-ci-hygiene` on `master`. Fixes #10107.

## Summary

PR #10024 added two different definitions of `is_rust_doctest_fence` in the same diff — two drafting passes of the same helper, neither removed. It merged while its CI run was cancelled, so E0428 never surfaced on a check.

Three commits:

- **be7a8b0** keeps the comma-splitting definition, deletes the whitespace-splitting one, and fixes the rustfmt and `clippy::question_mark` failures the same cancelled run hid in the same block.
- **7d0c90e** replaces the surviving whitelist with rustdoc's actual fence grammar, after a CodeRabbit review finding. See below.
- **98b030c** replaces the separator closure `7d0c90e` introduced with a spelled-out set, after `ripr+ New Gap Gate` flagged the closure as an unprovable seam. See below.

### Choosing between the two definitions

They differed only in how they split the fence info string:

| | separator | classifies `rust,ignore` |
| --- | --- | --- |
| line 2689 | whitespace | **not Rust** — skipped |
| line 2725 | comma | Rust — scanned |

The repository's own fences settle it. Of 1,022 fenced blocks under `crates/` and `xtask/`:

```
556  (bare)      126  rust        46  ignore      17  perl      2  rust,no_run
165  rust,ignore  47  text        34  bash        10  json      2  lua
                              19  no_run
```

The whitespace-splitting variant reads `rust,ignore` as one token, matches nothing, and silently skips all 167 comma-form Rust doctests — a scanner that reports clean because it never looked.

Measured against the committed baselines, both variants pass the ratchet; the kept one detects 5 more real hits:

```
comma-split (kept)         1,342 unwrap/expect   550 panic-family
whitespace-split (dropped) 1,337 unwrap/expect   550 panic-family
baselines                  1,528 unwrap/expect   568 panic-family
```

### Matching rustdoc's grammar (7d0c90e)

CodeRabbit pointed out that the surviving whitelist was still narrower than rustdoc. It was right. Rustdoc separates fence attributes with commas **or** whitespace, so `rust no_run` is as valid as `rust,no_run`, and an exact-match whitelist also missed `edition####`, `standalone_crate`, `test_harness`, and `ignore-<target>`.

Rather than lengthen the whitelist, the predicate now matches how rustdoc decides: split on both separators, and treat the block as a Rust doctest unless one of its attributes names something else. An unrecognized attribute (`text`, `bash`, `perl`, `json`) still disqualifies it, so the scanner does not begin reporting hits inside blocks rustdoc never compiles. The bare fence — 556 of the 1,022 — stays Rust.

This is ratchet-neutral by measurement rather than by argument: no fence in the tree uses any newly recognized form, so the counts are identical before and after (1,342 and 550).

It also removed an unreachable `else { return true }`, since `str::split` always yields at least one item.

Not implemented: the `ignore (reason)` parenthesized form also raised in the review. It does not appear in the rustdoc documentation — the documented reason form is `ignore-<target>`, which is handled — so recognizing it would widen detection on syntax rustdoc does not define.

### Making the separator set provable (98b030c)

`ripr+ New Gap Gate` went red on `7d0c90e` with one severe finding, and it was right. The grammar rewrite expressed the separator test as a closure:

```rust
.split(|character: char| character == ',' || character.is_whitespace())
```

RIPR classified that seam `weakly_exposed`. Its evidence path says exactly why:

```
reach yes:      Related tests appear to reach is_rust_doctest_fence
infection weak: Related tests contain input values, but the equality-boundary
                discriminator is missing
missing:        No related test call uses .split(|character: char| character
                equal to ','...); observed character values: unknown
```

The tests do drive that boundary — `is_rust_doctest_fence("rust,ignore")` hits it on the first character. RIPR cannot see it, because the closure's parameter is fed by `split`'s internals rather than by a test argument, so it cannot connect a literal in a test to the value at the boundary. This repository has hit the same class before in this same crate and answered it with a narrow suppression (`ripr-suppress-print-in-lib-allow-scopes-iterator-seams`, ripr#1429). A suppression turns the gate off for the file; it does not make the code any more provable, and a gate that is easier to switch off than to satisfy stops being a gate.

So the separators are spelled out as a set instead, with an explicit loop in place of `filter().all()`:

```rust
const FENCE_ATTRIBUTE_SEPARATORS: [char; 5] = [',', ' ', '\t', '\r', '\n'];
```

There is no predicate left for RIPR to find unreachable, and the boundary values are literals a test can name. No suppression added.

This narrows the separator set from every Unicode whitespace character to five. A fence info string is one source line, so space and tab are the only whitespace that can occur, plus a line ending on a CRLF checkout. Neutral by measurement again: 1,342 and 550, unchanged.

## Lane

- [ ] trust
- [x] substrate
- [ ] reliability

## Publication Sync

- Publication-sync PR (yes/no): no

## Claim Boundary

Claims that the workspace compiles on this head (`cargo check --workspace --all-targets --locked`, exit 0, warnings only), that `perl-ci-hygiene` additionally formats, lints under `-D warnings`, and tests clean, that the test-source ratchet's counts are unchanged by this PR, and that diff-scoped RIPR reports 0 severe gaps on this head with 0 findings suppressed by policy.

Does not claim the workspace passes `fmt`, `clippy`, or the full test suite — only `check`. Does not claim the ratchet's baselines are correct, only that this PR does not move the counts. Does not claim the fence predicate matches rustdoc exhaustively; it matches the documented attributes, and the separators that can occur in a single-line info string.

## Non-goals

- no change to the ratchet baselines
- no change in `perl-lsp-swarm`. The development repository modularised `perl-ci-hygiene` into `src/commands/panic_test/` and carries no copy of `is_rust_doctest_fence`, `scan_doctest_source`, or `rustdoc_line`, so there is nothing to sync in either direction
- does not address the merge-with-cancelled-CI path that let this land (#7278)
- does not fix the required check that cannot report (#10109), found while verifying CI here
- does not touch the expired Scenario 67 UX waivers (#10111), also found while verifying CI here
- does not renew the expired quality-gate exception review dates that keep `ripr+ New Gap Gate` red on every PR (#10112), found while verifying CI here

## Promotion Discipline

N/A — substrate lane.

## Behavior

- [x] no behavior change

The only runtime effect is on a CI hygiene scanner that has been uncompilable for five weeks, so there is no in-service behavior to change. The scanner's classification is now broader in principle, but provably identical on this tree.

## Risk Surfaces

None of the listed surfaces. The change is confined to a CI hygiene binary.

## Changes

`crates/perl-ci-hygiene/src/main.rs`, one file, three commits:

1. delete the whitespace-splitting `is_rust_doctest_fence` (the E0428 fix)
2. `clippy::question_mark` — `let Some(content) = ... else { return None }` becomes `?` in `rustdoc_line`, an error under `-D warnings`
3. rustfmt line wrapping in `scan_doctest_source` and `rustdoc_line`
4. `is_rust_doctest_fence` split into three documented functions matching rustdoc's grammar, plus two tests
5. the separator closure replaced by `FENCE_ATTRIBUTE_SEPARATORS` and an explicit loop, plus one test

## Test

Three tests added:

- `rust_fences_cover_rustdoc_attribute_forms` — every attribute form, both separators, negatives for the five other languages in the tree, a mixed `rust,text`, and `editionfoo`/`edition` to pin the four-digit shape
- `doctest_scanner_reports_hits_in_comma_form_rust_fences` — `rust,ignore` and `rust no_run` end to end through `scan_doctest_source`
- `rust_fence_separators_are_exactly_the_documented_set` — each separator driven individually, the repeated/leading/trailing empty-attribute cases, and `rust;no_run` / `rust|no_run` as the discriminators for the set itself: widening it would flip both to true

Worth recording why the first two were needed: the pre-existing `doctest_scanner_only_reports_rust_fences` used only bare `rust` and `text`, which **both** candidate definitions classified identically. It would have passed just as happily with the wrong one kept, so it could not have caught the choice going the other way.

The E0428 itself needs no regression test: a second definition of the same name cannot compile.

## Verification

- [x] Lane and risk surface are declared above.
- [x] Trust-lane PRs name promotion, fallback, blocker, and receipt boundaries. (N/A — substrate)
- [x] `cargo fmt -p perl-ci-hygiene -- --check` — clean
- [x] I used a narrow orthogonal pass first: reproduced E0428 directly, then surveyed the fence corpus and measured both variants against the baselines before running the broader gate.
- [x] `cargo clippy -p perl-ci-hygiene --all-targets --locked -- -D warnings` — clean
- [x] `cargo test -p perl-ci-hygiene --all-targets --locked` — pass
- [ ] This PR introduces UX-visible changes. — no

## Local Proof Commands

On `98b030c`:

- `cargo check -p perl-ci-hygiene --all-targets --locked` — pass (fails with E0428 on `master`)
- `cargo fmt -p perl-ci-hygiene -- --check` — pass (fails on `master`)
- `cargo clippy -p perl-ci-hygiene --all-targets --locked -- -D warnings` — pass (fails on `master`)
- `cargo test -p perl-ci-hygiene --all-targets --locked` — pass, 94 + 3 + 11 + 7 + 4
- `./target/debug/perl-ci-hygiene check-unwraps-tests` — pass, 1,342/1,528 and 550/568, unchanged across all three commits
- `cargo check --workspace --all-targets --locked` — pass, exit 0, warnings only
- `cargo install ripr --version 0.9.0 --locked` then `cargo xtask ripr-pr --base origin/master --head HEAD` — run on both heads:
  - on `7d0c90e`: `severe_gaps: 1`, `weakly_exposed: 1`, `ripr_severe_gap: true`, `requires_targeted_mutation: true`
  - on `98b030c`: `severe_gaps: 0`, `weakly_exposed: 0`, `reachable_unrevealed: 0`, `no_static_path: 0`, `suppressed_by_policy: 0`
- `cargo fmt --all -- --check` — not run
- `cargo clippy --workspace` — not run
- `cargo test --workspace` — not run

## Quality Gates

- new RIPR gaps: 0 on `98b030c`, measured with the same `cargo xtask ripr-pr` the workflow runs. `7d0c90e` had 1 severe gap; 98b030c closes it by making the seam traceable, not by suppressing it — `suppressed_by_policy` is 0 and `policy/ripr-suppressions.toml` is unchanged
- total RIPR+ gaps: unchanged
- patch coverage: the fence predicate and its two helpers are directly covered by the three tests above
- project coverage: unchanged
- receipt freshness: N/A
- exception status: none
- local verify command: `cargo test -p perl-ci-hygiene --all-targets --locked`
- receipt command: `./target/debug/perl-ci-hygiene check-unwraps-tests`

## RIPR / Coverage Effect

0 severe RIPR gaps on the current head, down from 1 on `7d0c90e`, with no suppression added; patch coverage unchanged; project coverage unchanged.

## Cleanup Performed

Removed the scratch worktree used to confirm the rustfmt and clippy failures pre-existed on a clean `origin/master` checkout, and restored the kept implementation after measuring the discarded one. A fixture modified while measuring the Scenario 67 waivers for #10111 was restored; that investigation contributed no change to this branch. Working tree contains only the changes above.

## Retained State

N/A.

## What I considered but didn't do

**Keeping the narrow whitelist.** I originally argued against restoring `standalone_crate`, `edition*`, and `ignore*` on the grounds that widening detection could push the ratchet up for no measured gain. That was an argument made where a measurement was available, and the measurement shows the counts do not move. 7d0c90e reverses it.

**A RIPR suppression for the closure seam.** The repository has an accepted precedent for exactly this class, on exactly this crate, citing ripr#1429. Adding a sixty-seventh suppression would have been the cheaper route and would have gone green just as fast. It would also have left the seam unprovable and the gate one file weaker. Rewriting the expression cost four lines.

**The `ignore (reason)` form.** Not documented rustdoc syntax; see above.

**Changing the baselines.** Not needed — measured, unchanged.

## Remaining Work

- the merge-with-cancelled-CI path that let a non-compiling diff reach `master` and stay there for five commits — #7278
- `Perl LSP Rust Small Result` cannot report on any `master` PR — #10109, found while verifying CI here
- `UX Regression Tests` has been red on every PR in this repository since 2026-08-01 — #10111, found while verifying CI here. Not this PR's; the dependency graph rules it out
- `ripr+ New Gap Gate`, a required check, is red on this head and on every other PR in this repository — #10112. `policy/quality-gate-exceptions.toml` sets `due_review = "fail"` with `review_after = "2026-07-28"` on both burn-down exceptions, so the gate raises two blocking actions on any commit before it reads a receipt. This branch's own diff-scoped RIPR evidence is clean at the same head. `perl-lsp-swarm` `origin/main` already carries the renewed file; porting it changes the posture of a branch-protection required check, so it is the maintainer's call and is not done here
- workspace `fmt`, `clippy`, and the full test suite are unverified; only `check` was run workspace-wide

## CI cost / verification note

- [x] I used the cheapest relevant proof first — direct `cargo check` on the one crate before anything broader.
- [x] I did not request broad CI beyond this PR's risk surface.
- [x] No high-cost CI label applied.
- [x] The gate failure was reproduced locally before pushing a fix for it, rather than spending a CI cycle to guess: `ripr` 0.9.0 installed locally, the workflow's own `cargo xtask ripr-pr` run on both heads, and the before/after recorded above.

## Agent

Claude Code, from a project thread. Controlling issue #10107.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_0122nfiQUfT7ojhCY2fYWuPc

---
_Generated by [Claude Code](https://claude.ai/code)_

### CODE REVIEW

**Comment by EffortlessSteven** [comment]
Addressed the 🟠 Major finding in 7d0c90e — it was correct, and my original reasoning for leaving it was wrong.

I had noted in the PR body that I was deliberately not restoring `standalone_crate`, `edition*`, and `ignore*`, on the grounds that widening detection could only push the ratchet up for no measured gain. That was an argument where a measurement was available. No fence in the tree uses any of those forms, so the counts cannot move:

```
before    1,342 unwrap/expect   550 panic-family
after     1,342 unwrap/expect   550 panic-family
baseline  1,528 unwrap/expect   568 panic-family
```

The finding is right on the grammar. Rustdoc separates fence attributes with commas **or** whitespace, so `rust no_run` is as valid as `rust,no_run`, and the comma-only split missed it along with `edition####`, `standalone_crate`, `test_harness`, and `ignore-<target>`.

Rather than extend the whitelist, I inverted the test to match how rustdoc actually decides. The fence is split on both separators, and the block is a Rust doctest unless one of its attributes names something else. An unrecognized attribute (`text`, `bash`, `perl`, `json`) still disqualifies it, so the scanner does not begin reporting hits inside blocks rustdoc never compiles. The bare fence — 556 of the 1,022 in the tree — stays Rust, since the attribute list is empty and `all` returns true.

This also removed the unreachable `else { return true }` I had flagged as follow-up: `str::split` always yields at least one item, so that arm was dead.

**One part not implemented:** the `ignore (reason)` parenthesized form. I could not find it in the rustdoc documentation — the documented reason form is `ignore-<target>`, which is now handled. Recognizing a form rustdoc does not define would widen detection on syntax that cannot appear. Happy to add it if you can point at where it is specified.

Two tests added, since the existing one could not have caught the wrong implementation being kept — it used only bare `rust` and `text`, which both candidate implementations classified identically:

- `rust_fences_cover_rustdoc_attribute_forms` — every attribute form above, plus negatives for the five other languages in the tree, a mixed `rust,text`, and `editionfoo`/`edition` to pin the four-digit shape
- `doctest_scanner_reports_hits_in_comma_form_rust_fences` — `rust,ignore` and `rust no_run` end to end through `scan_doctest_source`

Clippy under `-D warnings`, fmt, and the full crate test suite are clean on 7d0c90e.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EffortlessSteven** [comment]
## Codecov reports a failing `ux-regression` test — it is not from this PR

Recording why I am not acting on it, rather than leaving it silent.

Codecov's comment reports `ux-regression::ux-regression` failing, class `test_race`, route `test_fix`, 0s run time. No UX regression test ran on this branch, on either head:

- `UX Regression Tests` appears in the check runs for `be7a8b0` as **skipped**, and does not appear at all on `7d0c90e`
- across all 27 workflow runs on this branch the conclusions are 14 skipped, 8 success, 5 cancelled — **zero failures**
- the 5 cancelled runs are all on the superseded head `be7a8b0`, cancelled by `cancel-in-progress` when I pushed `7d0c90e`

So no job on this PR produced that result. The Codecov dashboard link is branch-scoped, and the 0s run time with no stack frames suggests aggregated or stale test-analytics data rather than a run on these commits.

On the merits it could not be this PR's either: the diff is 1 file, `crates/perl-ci-hygiene/src/main.rs`, a CI hygiene binary with no LSP, workspace-indexing, or UX surface. And `master` itself has not compiled since 4f0de95, so there is no green UX baseline on the base branch to regress from.

No fix to port, because there is no failing job here to fix. I have not re-run anything — re-running a job that never ran would prove nothing.

## Separately: one required check is not reporting, and cannot

While checking the above I found that `Perl LSP Rust Small Result` has not reported on this PR and structurally cannot. Its workflow triggers on `pull_request: branches: [main]`, and this repository has no `main` branch — `em-ci-routed-rust.yml` has **0 runs all time**, against 1,798 for `ripr.yml`, which lists `[master, main]`.

Filed as #10109. It is not this PR's to fix, and GitHub currently reports this PR as `unstable` rather than `blocked`, so it is not gating merge here.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EffortlessSteven** [comment]
## Correcting two things I said earlier, and the UX gate

### Correction 1 — the UX job has now run, and it failed

In [my earlier comment](https://github.com/EffortlessMetrics/perl-lsp/pull/10108#issuecomment-5746317576) I wrote that no UX regression test had run on this branch on either head, and that there was "no failing job here to fix." That was accurate at 00:11 and is not accurate now. `UX Regression Tests` ran on `7d0c90e` and failed at 00:21:40 ([job](https://github.com/EffortlessMetrics/perl-lsp/actions/runs/35477852790/job/105990625079)). The rest of that comment — the `Perl LSP Rust Small Result` finding, #10109 — stands.

### Correction 2 — this PR does introduce new RIPR gaps

The PR body's Quality Gates section says "new RIPR gaps: none — no production crate touched." The live gate disagrees, and the gate is right. `ripr+ New Gap Gate` failed on `7d0c90e` with three new unresolved gaps, all on seams `7d0c90e` created:

```
input that hits the boundary: character == ','
input that hits the boundary: matches!(attribute, "rust" | "no_run" | ...) || attribute == "ignore"
input that hits the boundary: attribute == "ignore"
```

That is mine and I am working it. The earlier run that showed `ripr` green on `be7a8b0` finished in 7 seconds — it was the draft skip, not a result.

---

## UX Regression Tests: root cause, and why I am not fixing it here

Five `error_waivers` in `crates/perl-lsp-ux-tests/fixtures/golden_editor_workload.json` all carry `"expires_after": "2026-08-01"`. `validate_waiver_expiry` rejects any waiver past its date, so `validate_manifest` fails before the workload runs. Reproduced on this head:

```
$ cargo test -p perl-lsp-ux-tests --test ux_scenario_67_golden_editor_workload -- --test-threads=1
test golden_manifest_has_exact_after_ready_shape ... FAILED
test scenario_67_golden_editor_workload_receipt ... FAILED
Error: error waiver expired on 2026-08-01; refresh or remove the waiver
test result: FAILED. 5 passed; 2 failed
```

**Not this PR's.** There is no dependency edge: `xtask` is the only workspace member that depends on `perl-ci-hygiene`, and neither `perl-lsp-ux-tests` nor the `perl-lsp` binary is downstream of it. The failure is a date comparison against `today()`, so it is deterministic on every commit in this repository on any date after 2026-08-01 — every PR here has had a red UX job for the last 50 days. I am not spending a re-run on it; a second run cannot change a date comparison.

Codecov's earlier comment reports this as `Failure class: test_race`, route `test_fix`. That classification does not match the cause. The scenario-67 receipt on my local reproduction of this head records `failure_class: unknown`, `route: triage`.

**Deleting the waivers is not the fix.** I measured it rather than assuming. With `"error_waivers": []` the manifest validates and `golden_manifest_has_exact_after_ready_shape` passes, but the workload still fails on the zero-budget check `unwaived request errors remain zero`, with `unwaived_error_count = 4` — one per project, all on the same journey:

```
mojolicious     | edit_burst_completion | -32800 Request superseded: document moved from generation 6 to 7 while textDocument/completion was queued
dancer2         | edit_burst_completion | -32800 Request superseded: generation 19 to 20
catalyst        | edit_burst_completion | -32800 Request superseded: generation 11 to 20
plain_modern_oo | edit_burst_completion | -32800 Request superseded: generation 19 to 20
```

So four of the five waivers are still load-bearing here. The fifth (`plain_modern_oo` / `rename_safe_lexical` / `timeout`) matched no error row in that run; one run is not enough to retire it.

**Where this gets interesting.** The waivers cite `"issue": 4050`. That is swarm [#4050](https://github.com/EffortlessMetrics/perl-lsp-swarm/issues/4050), closed 2026-08-12 by swarm [#7152](https://github.com/EffortlessMetrics/perl-lsp-swarm/pull/7152), "test(ux): disposition expired Scenario 67 waivers" — which set `error_waivers` to `[]`. Two things follow. The disposition never reached the publication repository. And the two fixtures had already diverged before it: swarm carried four waivers expiring 2026-08-11, this repository carries five expiring 2026-08-01. Since removing them here still leaves four live `request_superseded` errors, `perl-lsp` also appears to be missing whatever change let swarm's zero-waiver manifest pass.

That is a publication-sync question with a real behavioral component, not a fixture edit, and it is well outside a PR whose whole claim is that `master` compiles. Filed as #10111, with the measurements above and the two instrument problems that let it sit unread for 50 days.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EffortlessSteven** [comment]
`ripr+ New Gap Gate` is red on `98b030c`, and it is not this PR's failure. Filed as #10112.

The gate fails on a date, before it reads any receipt. `policy/quality-gate-exceptions.toml` sets `due_review = "fail"` and gives both burn-down exceptions `review_after = "2026-07-28"`; `quality_gate.rs:639` raises a `quality_exception_review_due` action once that date passes and `quality_gate.rs:1334` marks it blocking whenever `due_review != "warn"`. Reproduced locally on this exact head:

```
decision fail  head 98b030cba72a76172558db43163be53dd7b5f6c0
blocking  quality_exception_review_due  ripr-total-burndown        | review_after 2026-07-28 is on or before 2026-09-20
blocking  quality_exception_review_due  project-coverage-burndown  | review_after 2026-07-28 is on or before 2026-09-20
```

This branch's own RIPR evidence is clean at the same head: `target/ripr/pr/repo-exposure.json` gives `severe_gaps: 0`, `reachable_unrevealed: 0`, `no_static_path: 0`, `suppressed_by_policy: 0`, so `new_unresolved` is 0. Every non-cancelled `ripr.yml` run from 2026-08-11 to now concluded `failure`, on dependabot, docs, security and maintainer branches alike.

A fix exists in `perl-lsp-swarm` `origin/main`: the same file with `due_review = "warn"`, `review_after` renewed to `2026-09-16`, and a `falsifier` on each entry. I am not porting it here. It changes the posture of a branch-protection required check in the publication repository, and this PR's subject is a duplicate function definition in `perl-ci-hygiene` — that is a maintainer's call, and #10112 lays out the two shapes it could take. Say the word and I will port it into this branch.

The other red check, `UX Regression Tests`, is #10111, for the same reason on a different date. `CI Gate (Advisory Aggregate)` is a downstream echo of it; all seven `CI Gate shard` jobs, `Compile All Targets`, `LSP Memory Smoke`, `PR Smoke` and `Publication Sync Contract` are green on this head.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EffortlessSteven** [comment]
CodeRabbit's summary and both of its pre-merge warnings are pinned to `be7a8b0`, two commits behind this head — its own marker reads `change_assessment_commit:"be7a8b0f35f527ec959f1de0034a0ddaaa2a4906"`. Both points are already satisfied on `98b030c`:

**Linked Issues check** — "restore the existing `standalone_crate`, `edition*`, and `ignore*` recognition." Done in `7d0c90e`, which is the commit that acted on CodeRabbit's own earlier review. `is_rust_fence_attribute` matches `standalone_crate` and `test_harness` directly, `ignore` and `ignore-<target>` by prefix, and `is_edition_fence_attribute` accepts any `edition####` with a four-digit year so that an attribute merely beginning with `edition` is not read as one. `rust_fences_cover_rustdoc_attribute_forms` drives each of those forms, with `editionfoo` and bare `edition` as negatives.

**Docstring Coverage 0.00%, "analyzed 3 functions"** — measured on a head where the three functions it counted did not yet exist in their current form. On this head, `FENCE_ATTRIBUTE_SEPARATORS`, `is_rust_doctest_fence`, `is_rust_fence_attribute` and `is_edition_fence_attribute` each carry a doc comment covering the rustdoc grammar rule it implements and why.

No change needed. Flagging it because a reviewer looking at "❌ 2" on the summary card would otherwise reasonably pause.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by EffortlessSteven** [comment]
Ported swarm's renewed quality-gate exception ledger in `9f21caa`, at Steven's direction. This supersedes my previous comment and the PR body's Non-goals and Remaining Work entries for #10112, both of which said the port was deliberately not done here.

`policy/quality-gate-exceptions.toml` is now `perl-lsp-swarm` `origin/main`'s file byte for byte. It renews both burn-down entries with measured evidence, adds a `disposition` and a `falsifier` to each, moves `review_after` to `2026-09-16`, and sets `due_review = "warn"` with the rationale in a comment on the line above: *review dates create owner work; they must not change an unrelated PR verdict.*

Measured before and after on this branch with the workflow's own gate invocation:

```
before (98b030c)
  blocking=true   quality_exception_review_due  ripr-total-burndown        | review_after 2026-07-28 is on or before 2026-09-20
  blocking=true   quality_exception_review_due  project-coverage-burndown  | review_after 2026-07-28 is on or before 2026-09-20

after (9f21caa)
  blocking=false  quality_exception_review_due  ripr-total-burndown        | review_after 2026-09-16 is on or before 2026-09-20
  blocking=false  quality_exception_review_due  project-coverage-burndown  | review_after 2026-09-16 is on or before 2026-09-20
```

The two entries stay active and still appear in the receipt; they no longer fail the gate. What still blocks is unchanged: `expires = "2026-09-30"` blocks hard when it passes, and a new diff-scoped RIPR gap blocks via `new_unresolved`. This branch's own RIPR evidence remains clean at `severe_gaps: 0` with `suppressed_by_policy: 0`.

The evidence strings and the `#6216` / `#13704` references are swarm's own numbering, kept verbatim because this file is published from there rather than authored here. #10112 raised whether this should instead arrive through the publication-sync contract; it is filed there for that decision, and a later sync of this path is a no-op against this content.

Scope note: this PR's subject was a duplicate function definition in `perl-ci-hygiene`. The second file is here because `ripr+ New Gap Gate` is a required check and could not go green on any commit without it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/perl-ci-hygiene/src/main.rs [hunk 1]
```diff
@@ -2722,12 +2708,58 @@ fn strip_rust_comments(line: &str, in_block_comment: &mut bool) -> String {
...
+/// treated as an unrecognized language.
 fn is_rust_doctest_fence(fence: &str) -> bool {
-    let Some(language) = fence.trim().split(',').next() else {
-        return true;
-    };
-    language.is_empty()
-        || matches!(language, "rust" | "no_run" | "compile_fail" | "should_panic" | "ignore")
+    for attribute in fence.split(FENCE_ATTRIBUTE_SEPARATORS) {
+        if !attribute.is_empty() && !is_rust_fence_attribute(attribute) {
+            return false;
```
[the remark this answers is not in the document]
Comment (EffortlessSteven) [reply crates/perl-ci-hygiene/src/main.rs]: **Disposition: fixed** in `588b3d9`. Both halves of this were correct, and my PR body's claim that the parenthesized form "does not appear in the rustdoc documentation" was wrong.

**Evidence:** rather than settle it from the prose, I put one doctest per form in a scratch crate and ran `cargo test --doc`. `ignored` and `ok` both mean rustdoc collected the block; absence means it did not. Observed:

```
```ignore (needs network)       collected      scanner said skip   ← bug
```rust,text                    collected      scanner said skip   ← bug
```text,rust                    collected      scanner said skip   ← bug
```rust,perl                    collected      scanner said skip   ← bug
```ignore (why) no_run          collected      scanner said skip   ← bug
```ignore (one) (two)           collected      scanner said skip   ← bug
```(just a comment)             collected      scanner said skip   ← bug
```text  ```perl  ```wibble     not collected  scanner agreed
```(why) text                   not collected  scanner agreed
```

So two rules: parentheses enclose a comment and are stripped before the split (which also stops a comma inside a comment from splitting it), and an explicit `rust` anywhere wins over a foreign token in either order.

**Two things the probe settled that the suggestion didn't cover.** Comments do not nest — rustdoc warns on ```` ```ignore (outer (inner) text) ```` and declines to collect it. Scanning from one `(` to the next `)` reproduces that exactly, because the leftover `text)` stops qualifying, so no depth tracking is warranted and the simpler pass is also the accurate one. And the marker is case-sensitive: ```` ```Rust ````, ```` ```Rust,text ```` and ```` ```rustc ```` are none of them collected, which the implementation already got right and now has tests for.

**Not widened further.** The change is ratchet-neutral by measurement — 1342 unwrap/expect and 550 panic-family, unchanged — because no fence in this tree uses any newly recognized form. The widening is real but currently unexercised here.

Three tests added covering every row above, plus direct coverage of the comment-stripping and explicit-`rust` boundaries. `ripr+` measured locally at each step: 3 severe gaps on the first shape, 1 on the second, 0 on this one, with `suppressed_by_policy` at 0 throughout and `policy/ripr-suppressions.toml` untouched.

---
_Generated by [Claude Code](https://claude.ai/code)_
